### PR TITLE
override tab key for inline

### PR DESCRIPTION
**Pull request by dimacodota**
**motivation**: in the case where both a popup and the inline are visible, 'tab` key selects the popup suggestion:

**solution**: 
this pr adds the following : 
 - `lookahead` mode(the code was extracted on moved to a new file) - in this mode tabnine extends the selected  popup item and suggests a longer inline suggestion
 - the current `lookahead` suggestion is stored
 - we override the `tab` key and check if there is an active `lookahead` suggestion and accept it if any and fallback to the regular acceptance logic if there is no active `lookahead` suggestion
 
 
**before this fix**:

https://user-images.githubusercontent.com/60742964/178226600-da3a9a43-d2af-4432-8d58-bbf07f4389bb.mov 

**after this fix**:

https://user-images.githubusercontent.com/60742964/178226594-18e62b6c-e39c-461e-b175-ab89724d5735.mov

### CODE REVIEW

File: src/test/suite/completion.test.ts [hunk 1]
```diff
@@ -44,4 +61,53 @@ describe("Should do completion", () => {
...
+  expected: string,
+  doBeforeInline?: () => void
+) {
+  await openDocument("javascript", "cons");
+  await isProcessReadyForTest();
+  await moveToActivePosition();
+  await makeAChange("o");
+  await emulationUserInteraction();
+  await vscode.commands.executeCommand("editor.action.triggerSuggest");
+  doBeforeInline?.();
```
Comment (yonip23) [inline src/test/suite/completion.test.ts:104]: `?.()`
this is insane

File: src/test/suite/completion.test.ts [hunk 2]
```diff
@@ -44,4 +61,53 @@ describe("Should do completion", () => {
...
+  it("should prefer an inline when both popup and inline are visible", async () => {
+    await assertSuggestionWith("console.log", () => {
+      mockAutocomplete(
+        requestResponseItems,
+        anAutocompleteResponse("console", "console.log")
+      );
+    });
+  });
 });
+async function assertSuggestionWith(
```
Comment (yonip23) [inline src/test/suite/completion.test.ts:94]: too generic name, its not clear at all what does this function do. especially when there are tests which only calls this function and nothing else.

either separate it to more functions / rename it to be very descriptive

File: src/test/suite/completion.test.ts [hunk 3]
```diff
@@ -44,4 +61,53 @@ describe("Should do completion", () => {
...
+    await assertSuggestionWith("console");
+  });
+  it("should prefer an inline when both popup and inline are visible", async () => {
+    await assertSuggestionWith("console.log", () => {
+      mockAutocomplete(
+        requestResponseItems,
+        anAutocompleteResponse("console", "console.log")
+      );
+    });
+  });
```
Comment (yonip23) [inline src/test/suite/completion.test.ts:92]: its not clear what are these tests actually do. 
along with my other comment, i'd recommend extracting the logic to more functions

File: src/lookAheadSuggestion.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  );
+}
+async function enableTabOverrideContext(): Promise<Disposable> {
+  const res = await commands.executeCommand(
+    "setContext",
+    "tabnine.tab-override",
+    true
+  );
+
+  console.log("res after enableTabOverrideContext: ", res);
```
Comment (yonip23) [inline src/lookAheadSuggestion.ts:107]: do you need this in production? if not, remove

File: src/lookAheadSuggestion.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    (_textEditor: TextEditor, edit: TextEditorEdit) => {
+      const { range, insertText } = currentLookAheadSuggestion ?? {
+        range: undefined,
+        insertTex: undefined,
+      };
+      if (range && insertText) {
+        edit.replace(range, insertText);
+      } else {
+        void commands.executeCommand("acceptSelectedSuggestion");
+      }
```
Comment (yonip23) [inline src/lookAheadSuggestion.ts:96]: i think its more straight forward to write it something like:

```typescript
if (!currentLookAheadSuggestion) {
    void commands.executeCommand("acceptSelectedSuggestion");
    return;
}

const { range, insertText } = currentLookAheadSuggestion;
edit.replace(range, insertText);
```

File: package.json [hunk 6]
```diff
@@ -335,6 +335,11 @@
 				"command": "tabnine.accept-inline-suggestion",
 				"when": "tabnine.snippet-suggestion:enabled && tabnine.in-inline-suggestions || tabnine.inline-suggestion:enabled && tabnine.in-inline-suggestions"
 			},
+			{
+				"key": "tab",
+				"command": "tabnine.tab-override",
+				"when": "tabnine.tab-override && suggestWidgetVisible && textInputFocus"
+			},
```
Comment (yonip23) [inline package.json:342]: how does it play with the other "tab" key we're defining? can there be a situation where both key overrides are active? what happens if so?

**Review by yonip23 — CHANGES_REQUESTED** [review]
im requesting changes mainly for the tests, other than that great work 🙂

File: package.json (hunk 6 above)
Comment (dimacodota) [reply package.json:342]: each override should be guarded with the relevant context flag. this override is enabled only if `tabnine.tab-override` context flag is enabled.

**Pull request merged**